src/ripple/app/ledger/impl/TransactionMaster.cpp

```diff
@@ -54,44 +54,57 @@ TransactionMaster::fetch_from_cache(uint256 const& txnID)
     return mCache.fetch(txnID);
 }
 
-std::shared_ptr<Transaction>
+std::optional<std::pair<std::shared_ptr<Transaction>, std::shared_ptr<TxMeta>>>
 TransactionMaster::fetch(uint256 const& txnID, error_code_i& ec)
 {
-    auto txn = fetch_from_cache(txnID);
+    // txn->getLedger() == 0 then txn is not validated and does not contain
+    // metadata
+    if (auto txn = fetch_from_cache(txnID); txn && txn->getLedger() == 0)
+        return std::pair{std::move(txn), nullptr};
 
-    if (txn)
-        return txn;
+    auto v = Transaction::load(txnID, mApp, ec);
 
-    txn = Transaction::load(txnID, mApp, ec);
+    if (!v)
+        return std::nullopt;
 
-    if (!txn)
-        return txn;
+    auto [txn, txnMeta] = v.value();
 
-    mCache.canonicalize_replace_client(txnID, txn);
+    if (txn)
+        mCache.canonicalize_replace_client(txnID, txn);
 
-    return txn;
+    return std::pair{std::move(txn), std::move(txnMeta)};
 }
 
-boost::variant<Transaction::pointer, bool>
+std::optional<std::variant<
+    std::pair<std::shared_ptr<Transaction>, std::shared_ptr<TxMeta>>,
+    bool>>
```

Comment:
Wow, that's a complex return type: an optional variant of a pair and a bool?!? Yuck.

 I don't think it's worth holding up this PR, but let's make a note to refactor this in the near future. This return type needs to be simplified. One idea is to return a variant of the pair and some enum.

Reply:
Been thinking about it and I think the way to go here is to have a return type of 
```
std::variant<std::pair<std::shared_ptr<Transaction>, std::shared_ptr<TxMeta>>, SearchedAll>
```
given that the `optional` is representing a `SearchedAll::Unkown`. Thoughts?

Reply:
I think we should use `std::monostate` here: https://en.cppreference.com/w/cpp/utility/variant/monostate